### PR TITLE
[GLUTEN-3553][CH] Support bucket scan for ch backend

### LINKED ISSUE #3553: [CH] Support bucket scan for ch backend
### Description

Support bucket scan for ch backend, including parquet format and mergetree format.

### PR DESCRIPTION
**Pull request by zzcclp**
## What changes were proposed in this pull request?

Support bucket scan for ch backend, including parquet format and mergetree format.

Close #3553.

(Fixes: #3553)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala [hunk 1]
```diff
@@ -219,27 +219,29 @@ trait HashJoinLikeExecTransformer
...
         case _ =>
           val readRel = RelBuilder.makeReadRel(
             new JArrayList[Attribute](plan.output.asJava),
             substraitContext,
             -1
           ) /* A special handling in Join to delay the rel registration. */
-          (readRel, plan.output, true)
+          // Make sure create a new read relId for the stream side first
+          // before the one of the build side, when there is not shuffle on the build side
+          (readRel, plan.output, true, substraitContext.nextRelId())
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:235]: @rui-mo  @PHILO-HE  please help to check, this change will make sure that it creates a new relId for the stream side before the build side, when the stream side is reading data from iterator and the build side is no shuffle.

File: gluten-core/src/main/scala/io/glutenproject/utils/InputPartitionsUtil.scala [hunk 2]
```diff
@@ -29,24 +29,12 @@ case class InputPartitionsUtil(
...
-      relation.sparkSession.sessionState.conf.bucketingEnabled && relation.bucketSpec.isDefined
-      && !disableBucketedScan
-    ) {
-      val spec = relation.bucketSpec.get
-      val bucketColumns = spec.bucketColumnNames.flatMap(n => toAttribute(n))
-      bucketColumns.size == spec.bucketColumnNames.size
-    } else {
-      false
-    }
-  }
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/utils/InputPartitionsUtil.scala:48]: @WangGuangxin  please help to review, thanks.

**Comment by PHILO-HE** [comment]
I note some code in gluten-core is changed, but CI jobs for velox backend are not triggered. It may be github action's bug that can happen for PR with many new files.

**Comment by zzcclp** [comment]
> I note some code in gluten-core is changed, but CI jobs for velox backend are not triggered. It may be github action's bug that can happen for PR with many new files.

How to fix this problem, or how to trigger velox ci manually ?

**Comment by PHILO-HE** [comment]
Maybe, you can try to temporarily remove the paths filter:
https://github.com/oap-project/gluten/blob/16e2874502d0f794d3a79bf44875fcc2ca1c3c0d/.github/workflows/velox_be.yml#L20

**Comment by zzcclp** [comment]
> Maybe, you can try to temporarily remove the paths filter:
> 
> https://github.com/oap-project/gluten/blob/16e2874502d0f794d3a79bf44875fcc2ca1c3c0d/.github/workflows/velox_be.yml#L20

It works, thanks.

File: .github/workflows/velox_be.yml [hunk 3]
```diff
@@ -17,25 +17,6 @@ name: Velox backend
...
-      - 'tools/gluten-it/**'
-      - 'tools/gluten-te/**'
-      - 'ep/build-arrow/**'
-      - 'ep/build-velox/**'
-      - 'cpp/*'
-      - 'cpp/CMake/**'
-      - 'cpp/velox/**'
-      - 'cpp/core/**'
-      - 'dev**'
-#      - 'substrait/substrait-spark/**'
```
Comment (zzcclp) [inline .github/workflows/velox_be.yml:38]: will revert after ci passed

**Comment by zhouyuan** [comment]
May be due to the old github action checkout used: `actions/checkout@v2`. Will update this in following commits.

> > I note some code in gluten-core is changed, but CI jobs for velox backend are not triggered. It may be github action's bug that can happen for PR with many new files.
> 
> How to fix this problem, or how to trigger velox ci manually ?

File: gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala (hunk 1 above)
Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:235]: When there is a shuffle in the stream side and no shuffle in the build side, the operation orders passed to the backend are: 0: `stream side read from iter`, 1:`build side scan`, 2: `build side filter`, 4: `join`, but the orders created in the join transformer are: 0:`build side scan`, 1: `build side filter`, 2: `stream side read from iter`, 4: `join`, because `stream side read from iter` will be registed after creating the build side, so they are different.

File: gluten-core/src/main/scala/io/glutenproject/utils/InputPartitionsUtil.scala (hunk 2 above)
Comment (WangGuangxin) [reply gluten-core/src/main/scala/io/glutenproject/utils/InputPartitionsUtil.scala:48]: LGTM

**Comment by zzcclp** [comment]
@PHILO-HE @rui-mo please help to check why all velox ci were failed.

**Review by PHILO-HE — DISMISSED** [review]
It looks this PR can increase ~70MB repo size due to the introduced binary files for test. Maybe, in the future, we can build a dedicated test repo for gluten and let it keep such binary files. cc @FelixYBW @zhouyuan
It's OK to me to merge this PR firstly. Thanks!

File: .github/workflows/velox_be.yml (hunk 3 above)
Comment (rui-mo) [inline .github/workflows/velox_be.yml:38]: Do you need to revert these changes?

Comment (zzcclp) [reply .github/workflows/velox_be.yml:38]: > Do you need to revert these changes?

yes, it needs.

Comment (zzcclp) [reply .github/workflows/velox_be.yml]: Done

**Review by rui-mo — APPROVED** [review]

**Review by baibaichen — APPROVED** [review]
LGTM
